### PR TITLE
Upgrade to Pydantic v2

### PR DESCRIPTION
**Pull request by jaykv**
This will have to be tested more thoroughly of course, but I hope this is a good start to move towards pydantic v2 :)

Changes:
- New `BaseModel` method names, decorators, and config dict:
  * `model.parse_obj` -> `model.model_validate`
  * `model.dict` -> `model.model_dump(mode="json")`
  * `@root_validator` -> `@model_validator`
  * `@validator` -> `@field_validator`
  * `parse_raw(raw_stuff)` -> `model_validate(json.loads(raw_stuff))`
  * `schema` -> `model_json_schema`
  * `json` -> `model_dump_json`
  * `class Config: ...` -> `model_config = ConfigDict(...)`
- `BaseModel` subclasses using `__root__` switched to subclass `RootModel` with `root` annotation
- New fields for `ValidationErrorElement`: `input` and `url`
- Updated custom type definition for `BaseFile` ([thanks to flask-openapi3's implementation](https://github.com/luolingchun/flask-openapi3/blob/v3.x/flask_openapi3/models/file.py#L11))
- Response on `ValidationError` updated to use `err.json(indent=2)` instead of `jsonify(err.errors())` to avoid `TypeError: _ is not JSONSerializable` errors from the new `input` field (`indent=2` to keep response consistent to current)
- Update model schema definitions key from `definitions` -> `$defs`
- `req_validation_error.model.__name__` -> `req_validation_error.title`
- `pydantic.EmailStr.validate` -> `pydantic.validate_email`

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -22,7 +22,8 @@ classifiers = [
     "Topic :: Software Development :: Libraries :: Python Modules",
 ]
 dependencies = [
-    "pydantic>=1.9.1",
+    "pydantic>=2.0",
+    "pydantic-settings"
```
Comment (yedpodtrzitko) [inline pyproject.toml:26]:
```suggestion
    "pydantic-settings",
```

File: setup.py [hunk 2]
```diff
@@ -6,10 +6,11 @@
 setup(
     name="spectree",
     install_requires=[
-        "pydantic>=1.2",
+        "pydantic>=2.0",
+        "pydantic-settings"
```
Comment (yedpodtrzitko) [inline setup.py:10]:
```suggestion
        "pydantic-settings",
```

**Review by kemingy — COMMENTED** [review]
Thanks for your contribution.

Pydantic v2 has too many breaking changes that lots of 3rd libraries are not able to upgrade in a short time.

This PR is not backward compatible (due to pydantic breaking changes). I'm not sure if we should make this library compatible with both pydantic v1 & v2. Or maybe we should add another abstract layer so that we can make it possible to support pydantic v1 & v2 and msgspec (ref #307)

File: spectree/spec.py [hunk 3]
```diff
@@ -335,12 +335,15 @@ def _generate_spec(self) -> Dict[str, Any]:
 
         if self.config.servers:
             spec["servers"] = [
-                server.dict(exclude_none=True) for server in self.config.servers
+                server.model_dump(exclude_none=True, mode="json")
```
Comment (yedpodtrzitko) [inline spectree/spec.py:338]: previously `.dict()` was returning a Python dictionary, now `.model_dump(mode="json")` returns json. Is this change intentional and if so, what difference does it make (I'm not familiar with this part of codebase, so I'd really like to know)?

File: tests/test_plugin_starlette.py [hunk 4]
```diff
@@ -365,7 +365,7 @@ def test_starlette_no_response(client):
     resp = client.get("/api/no_response")
     assert resp.status_code == 200, resp.text
 
-    resp = client.post("/api/no_response", json={"name": "starlette", "limit": 1})
+    resp = client.post("/api/no_response", json={"name": "starlette", "limit": "1"})
```
Comment (yedpodtrzitko) [inline tests/test_plugin_starlette.py:368]: why has this changed?

**Comment by yedpodtrzitko** [comment]
That's a huge contribution, your effort is really appreciated. 

Pydantic v2 however still have some bugs and the ecosystem hasnt adapted yet, so I think it's too early to merge this at this moment.

File: spectree/spec.py (hunk 3 above)
Comment (jaykv) [reply spectree/spec.py:338]: `mode="json"` just serializes the values for the returned dict to make sure there's no python objects so when it's jsonified for `openapi.json`, it won't cause any `_ is not JSONSerializable` errors to pop up when accessing apidocs. I noticed it happening with some openapi config models in general when running the falcon_demo.

in this case though and probably elsewhere, I might have overused it since `Server` model doesn't have any special types

File: tests/test_plugin_starlette.py (hunk 4 above)
Comment (jaykv) [reply tests/test_plugin_starlette.py:368]: Yeah this one was weird. So `/api/no_response` has this decorator `@api.validate(json=StrDict, resp=Response(HTTP_200=None),)` and `StrDict` model has annotation `root: Dict[str, str]`. With v1, when we passed `"limit": 1`, `parse_obj` would internally coerce that into a string, but now `model_validate` is more strict about coercing it seems like

**Comment by jaykv** [comment]
I understand of course, this is definitely just here to get the conversation started and test the waters. 

So is the ideal separation of layers for future spectree this?
1. Request/response model validation (pydantic v1/v2, msgspec)
2. OpenAPI schema generation
3. Route parsing, generation and validation integration into backends (flask, falcon, starlette, quart)
4. Apidocs (swagger, redoc)

**Comment by kemingy** [comment]
> I understand of course, this is definitely just here to get the conversation started and test the waters.
> 
> So is the ideal separation of layers for future spectree this?
> 
>     1. Request/response model validation (pydantic v1/v2, msgspec)
> 
>     2. OpenAPI schema generation
> 
>     3. Route parsing, generation and validation integration into backends (flask, falcon, starlette, quart)
> 
>     4. Apidocs (swagger, redoc)

Yeah. I wonder if we can detect the model (pydantic v1/v2, megspec) automatically. This layer should be used to validate the request/response and generate the OpenAPI model schema.

**Comment by AndreiDrang** [comment]
Hi
Perhaps it makes sense to just make a separate library? Smth like `spectreeV2`?

**Comment by kemingy** [comment]
> Hi Perhaps it makes sense to just make a separate library? Smth like `spectreeV2`?

You can use `pip install git+https://github.com/0b01001001/spectree.git@refs/pull/319/head` directly.

**Comment by DanCardin** [comment]
At least for v2, pydantic appears to be including a `v1` backport under `pydantic.v1` inside the v2 versions. That may or may not be useful in terms of converting internal usage or making it cross-compatible.

But it **seems** like it might make it possible for the library at current moment to "simply" alias all of its imports to pydantic to the v1 version. Such that I, as a user of pydantic (that doesn't necessarily stop at API models) could upgrade to v2 without impacting the spectree-compatible v1 API models.

**Comment by kemingy** [comment]
> At least for v2, pydantic appears to be including a `v1` backport under `pydantic.v1` inside the v2 versions. That may or may not be useful in terms of converting internal usage or making it cross-compatible.
> 
> But it **seems** like it might make it possible for the library at current moment to "simply" alias all of its imports to pydantic to the v1 version. Such that I, as a user of pydantic (that doesn't necessarily stop at API models) could upgrade to v2 without impacting the spectree-compatible v1 API models.

This sounds like a good temporary solution.

**Comment by jaykv** [comment]
> At least for v2, pydantic appears to be including a `v1` backport under `pydantic.v1` inside the v2 versions. That may or may not be useful in terms of converting internal usage or making it cross-compatible.
> 
> But it **seems** like it might make it possible for the library at current moment to "simply" alias all of its imports to pydantic to the v1 version. Such that I, as a user of pydantic (that doesn't necessarily stop at API models) could upgrade to v2 without impacting the spectree-compatible v1 API models.

I've seen this approach used specifically with just:
```python
try:
    from pydantic import v1 as pydantic
except ImportError:
    import pydantic
```

I didn't explore this yet, but could be worth looking into since it's a smaller change and also less likely to break for current v1 users. The only downside is spectree internally would still be using the v1 API instead of the v2 API. Although, that might not even matter since the long-term plan is to rehaul the internals to support msgspec anyways.

#324 is exploring the reverse approach for fun and upgrading the code to v2 API and creating backports for v1 and version-specific logic to handle the breaking changes.
